update_workbook.py

```diff
@@ -0,0 +1,249 @@
+    materials = []
+    for lecture_dir in lecture_to_directories.values():
+        try:
+            file_paths = os.listdir(root_dir/lecture_dir)
+            materials.extend([lecture_dir/Path(file_path) for file_path in file_paths])
+        except FileNotFoundError:
+            ...
+    
+    return materials
+
+
+def get_all_materials() -> dict:
+    textbook_list = listmaterials(root_dir=Path('./'))
+    workbook_list = listmaterials(root_dir=WORKBOOK_ROOT_DIR)
+
+    materials = dict()
+    for workbook in workbook_list:
+        if workbook not in textbook_list:
+            materials[workbook] = GitStatus.DELETED
+
+    for textbook in textbook_list:
+        materials[textbook] = GitStatus.TOUCHED
+
+    return materials
+
+
+def export_python_material(textbook_path: Path) -> None:
+    """Copy python materials to the workbook directory."""
+    import shutil
+    shutil.copy(textbook_path, WORKBOOK_ROOT_DIR/textbook_path.parent)
+
+
+# Helper - Jupyter Contents
+def retrieve_tags(cell: dict[str, dict]) -> list:
+    return meta.get('tags', []) if (meta:=cell.get('metadata')) else []
+
+
+def tags_in(cell: dict[str, dict], *tags: str) -> bool:
+    return (cell_tags:=retrieve_tags(cell)) is not None and \
+            all([tag in cell_tags for tag in tags])
+
+
+def is_solution_cell(cell: dict[str, dict]) -> bool:
+    return tags_in(cell, 'solution')
+
+
+def _get_or_make(obj: dict, key: str) -> dict:
+    if key not in obj:
+        obj[key] = dict()
+        return obj[key]
+    return obj[key]
+
+
+def mkdict(obj: dict, key: str, /, *args: str) -> dict:
+    child_obj = _get_or_make(obj, key)
+    if not args:
+        return child_obj
+    else:
+        return mkdict(child_obj, *args)
+
+
+class Textbook:
+    """Jpyter notebook for lectures."""
+    def __init__(self, rel_path: Path) -> None:
+        self.rel_path = rel_path
+        self.contents = self._load_contents()
+        self.workbook_path = WORKBOOK_ROOT_DIR/self.rel_path
+        self.workbook = self._create_workbook_contents()
+
+    def _create_workbook_contents(self) -> dict:
+        """
+        Update workbook(ipynb) from the textbook contents.
+
+        How to create a workbook from a textbook
+        ----------------------------------------
+        For all cells with ``solution`` tag,
+        remove ``source`` of the cell and ``hide-cell``, ``solution`` tags and
+        add ``SOURCE_REPLACEMENTM_MESSAE`` to the ``source``.
+        """
+        from copy import deepcopy
+        workbook = deepcopy(self.contents)
+        solution_cells = filter(is_solution_cell, workbook.get('cells', []))
+        
+        for solution in solution_cells:
+            solution['source'] = [SOURCE_REPLACEMENT_MESSAGE]
+            tags = retrieve_tags(solution)
+            for tag in ('hide-cell', 'solution'):
+                if tag in tags:
+                    tags.remove(tag)
+            if 'workbook' not in tags:
+                tags.append('workbook')
```

Comment:
What is this used for?

Reply:
I wanted to include metadata with the information of the original textbook and then I decided not to,
but I thought it would be better to keep an indicator that it is a modified cell.